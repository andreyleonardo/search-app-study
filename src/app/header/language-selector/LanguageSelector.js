import React, { Component } from 'react';
import PropTypes from 'prop-types';
import LanguageOptions from './language-options/LanguageOptions';

import './LanguageSelector.scss';

class LanguageSelector extends Component {
  constructor(props) {
    super(props);

    this.state = {
      displayOptions: false
    };

    this.onDisplayChange = this.onDisplayChange.bind(this);
    this.handleOnClick = this.handleOnClick.bind(this);
  }

  onDisplayChange() {
    this.setState(state => ({ displayOptions: !state.displayOptions }));
  }

  handleOnClick(language) {
    const { onClick, selectedLanguage } = this.props;
    this.onDisplayChange();
    if (language !== selectedLanguage) {
      onClick(language);
    }
  }

  render() {
    const { selectedLanguage } = this.props;
    const { displayOptions } = this.state;

    return (
      <div className="language_selector">
        <div className="language_selector__button">
          { !displayOptions ? (
            <button id="language-selector-id" type="button" onClick={this.onDisplayChange}>{ selectedLanguage }</button>
          ) : (
            <LanguageOptions onClick={this.handleOnClick} />
          )}
        </div>
      </div>
    );
  }
}

LanguageSelector.propTypes = {
  onClick: PropTypes.func.isRequired,
  selectedLanguage: PropTypes.string.isRequired
};

export default LanguageSelector;
