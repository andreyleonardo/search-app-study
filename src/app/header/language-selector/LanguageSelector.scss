@import "src/stylesheets/css/_colors";

.language_selector {
  &__button {
    display: flex;
    justify-content: flex-end;
  }

  button {
    border-radius: 3px;
    text-transform: uppercase;
    outline: none;
    letter-spacing: 0.5px;
    font-size: 1em;
    margin: 10px;
    border: none;
    background-color: white;
    font-size: 1em;
    cursor: pointer;
    // color: $primary;

    &.primary {
      background-color: $primary;
      color: white;
      &:hover {
        background-color: $secondary;
      }
    }

    &.secondary {
      background-color: $secondary;
      color: white;

      &:hover {
        background-color: $primary;
      }
    }
  }
}
