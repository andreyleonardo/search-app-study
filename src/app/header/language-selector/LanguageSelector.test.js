import React from 'react';
import LanguageSelector from './LanguageSelector';

describe('Language Selector', () => {
  let languageSelector;
  let props;

  beforeEach(() => {
    props = {
      onClick: jest.fn(),
      selectedLanguage: 'pt-BR'
    };
    languageSelector = shallow(<LanguageSelector {...props} />);
  });

  it('renders without break', () => {
    expect(languageSelector).toMatchSnapshot();
  });

  it('displays language options and hides language selector when it is clicked', () => {
    languageSelector.find('#language-selector-id').simulate('click');

    expect(languageSelector.exists('#language-selector-id')).toBeFalsy();
    expect(languageSelector.exists('LanguageOptions')).toBeTruthy();
  });

  it('calls onClick when a language is different from selectedLanguage', () => {
    languageSelector.setState({ displayOptions: true });

    languageSelector.instance().handleOnClick('en-US');

    expect(props.onClick).toHaveBeenCalledTimes(1);
  });

  it('does not calls onClick when a language is equal to selectedLanguage', () => {
    languageSelector.setState({ displayOptions: true });

    languageSelector.instance().handleOnClick('pt-BR');

    expect(props.onClick).not.toHaveBeenCalled();
  });

  it('closes language options and shows language selector button when a language is selected', () => {
    languageSelector.setState({ displayOptions: true });

    languageSelector.instance().handleOnClick('en-US');

    expect(languageSelector.exists('#language-selector-id')).toBeTruthy();
    expect(languageSelector.exists('LanguageOptions')).toBeFalsy();
  });

  it('hides language options when clicked outside', () => {
    languageSelector.setState({ displayOptions: true });
    languageSelector.update();

    expect(languageSelector.exists('LanguageOptions')).toBeTruthy();

    languageSelector.find('OutsideClickHandler').props().onOutsideClick();
    languageSelector.update();

    expect(languageSelector.exists('LanguageOptions')).toBeFalsy();
  });
});
