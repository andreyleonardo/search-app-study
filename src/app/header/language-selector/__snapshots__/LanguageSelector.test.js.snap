// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language Selector renders without break 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LanguageSelector
    onClick={[MockFunction]}
    selectedLanguage="pt-BR"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="language_selector__button"
      >
        <button
          id="language-selector-id"
          onClick={[Function]}
          type="button"
        >
          pt-BR
        </button>
      </div>,
      "className": "language_selector",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <button
          id="language-selector-id"
          onClick={[Function]}
          type="button"
        >
          pt-BR
        </button>,
        "className": "language_selector__button",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "pt-BR",
          "id": "language-selector-id",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": "pt-BR",
        "type": "button",
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="language_selector__button"
        >
          <button
            id="language-selector-id"
            onClick={[Function]}
            type="button"
          >
            pt-BR
          </button>
        </div>,
        "className": "language_selector",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            id="language-selector-id"
            onClick={[Function]}
            type="button"
          >
            pt-BR
          </button>,
          "className": "language_selector__button",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "pt-BR",
            "id": "language-selector-id",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "pt-BR",
          "type": "button",
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
